vg/vgl/canvas.go

```diff
@@ -0,0 +1,348 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package vgl
+
+import (
+	"fmt"
+	"image/color"
+	"runtime"
+	"time"
+
+	"github.com/go-gl/gl/v2.1/gl"
+	"github.com/go-gl/glfw/v3.1/glfw"
+	"github.com/gonum/plot/vg"
+	"github.com/llgcode/draw2d"
+	"github.com/llgcode/draw2d/draw2dgl"
+)
+
+// dpi is the number of dots per inch.
+const dpi = 96
+
+// Canvas implements the vg.Canvas interface, using OpenGL as a backend.
+type Canvas struct {
+	gc    draw2d.GraphicContext
+	win   *glfw.Window
+	w, h  vg.Length // width and height of the canvas
+	width vg.Length // width of the current line
+	color []color.Color
+}
+
+// make sure we implement vg.Canvas
+var _ vg.Canvas = (*Canvas)(nil)
+
+// channel where all OpenGL calls are made
+var glchan chan func()
+var done chan struct{}
+
+func init() {
+	// OpenGL calls need to be run on the main OS-thread.
+	// calling LockOSThread during init() ensures that is the case.
+	runtime.LockOSThread()
+}
+
+// Run runs OpenGL commands.
+// Run needs to be run on the main OS-thread.
+func Run(f func() error) error {
```

Comment:
What about an interface like this:
`NewWindow(w, h int, title string) Window`
`Window.Draw(f func(Canvas)error) error`
This way, you can draw multiple times to the same window.
Also, it's more clear that `Draw` expects `f` to draw to the given canvas.
With `Run`, it's not obvious from the API that you are expected to create a canvas to which you draw within `f`.